README.md
elixir-bson
===========
[![Build Status](https://travis-ci.org/checkiz/elixir-bson.png?branch=master)](https://travis-ci.org/checkiz/elixir-bson)
[![Hex Version](https://img.shields.io/hexpm/v/bson.svg)](https://hex.pm/packages/bson)
[![Hex Downloads](https://img.shields.io/hexpm/dt/bson.svg)](https://hex.pm/packages/bson)

BSON implementation for Elixir Language

elixir-bson on GitHub [source repo](https://github.com/checkiz/elixir-bson) -
[documentation](https://checkiz.github.io/elixir-bson)


BSON is a binary format in which zero or more key/value pairs are stored as a single entity, called a document. It is a data type with a standard binary representation defined at <http://www.bsonspec.org>.

This implements version 1.0 of that spec.

This project is used by [elixir-mongo](https://github.com/checkiz/elixir-mongo), a [MongoDB](http://www.mongodb.org) driver in Elixir.

This implementation maps the Bson grammar with Elixir terms in the following way:

  - document: Map, HasDict, Keyword
  - int32 and int64: Integer
  - double: Float
  - string: String
  - Array: List (non-keyword)
  - binary: Bson.Bin (struct)
  - ObjectId: Bson.ObjectId (struct)
  - Boolean: true or false (Atom)
  - UTC datetime: Bson.UTC (struct)
  - Null value: nil (Atom)
  - Regular expression: Bson.Regex (struct)
  - JavaScript: Bson.JS (struct)
  - Timestamp: Bson.Timestamp (struct)
  - Min and Max key: `MIN_KEY` or `MAX_KEY` (Atom)

This is how to encode a sample Elixir Map into a Bson Document:

```elixir
bson = Bson.encode %{a: 1, b: "2", c: [1,2,3], d: %{d1: 10, d2: 11} }

```
In this case, `bson` would be a document with 4 elements (an Integer, a String, an Array and an embeded document). This document would correspond in Javascript to:
```javascript
{a: 1, b: "2", c: [1,2,3], d: {d1: 10, d2: 11} }
```

Conversly, to decode a bson document:
```elixir
%{a: 1} == Bson.decode <<12, 0, 0, 0, 16, 97, 0, 1, 0, 0, 0, 0>>
```

Special Bson element that do not have obvious corresponding type in Elixir are represented with Record, for example:

```elixir
jsbson = Bson.encode js: %Bson.JS{code:"function(a) return a+b;", scope: [b: 2]}
rebson = Bson.encode re: %Bson.Regex{pattern: "\d*", opts: "g"}
```

Some configuration can be done using fun or protocol implementation, ie, it is possible to redefine encoder end decoder of Bson.Bin to implement specific encoding. For that you can set Application envir for application `:bson`. Two options are available: `:decoder_new_doc` defaulted to `Bson.Decoder.elist_to_atom_map/1` and `:decoder_new_bin` defaulted to `&Bson.Bin.new/2`.


CHANGELOG.md
# v0.4.4
* fix compilation warning
* fix documentation url
* add improvement from X4lldux
# v0.4.3
* Add config for decoder: decoder_new_doc and decoder_new_bin 
* Add %Bson.UTC inspect : when inspecting %Bson.UTC, convert it to ISO8601 format, e.g. 2014-9-11T22:13:54
* Decode crash if the buffer size is < 5, now it generates a %Bson.Decoder.Error{} return message
# v0.4.2
* new helper function `Bson.ObjectId.from_string/1`
# v0.4.1
* allow setting options to Bson.decode
* improve documentation
# v0.4.0
* prépartion pour elixir v1
* Bson.Regex is obsolete and replaced with Regex
* Improved error messages
* Keys of maps are strings when decoded, by default
* encoder prototcol does not fallback to any to allow custom implementation
* removed encoding / decoding of the triplet tuple representing a time stamp (need to implement locally if needed)
* UTC is implemented using struct `Bson.UTC`
* in `Bson.Bin`, subtypes are positif integer (<256) not a binary anymore
* all atom representing constant like :min_key are in small caps (used to be MIN_KEY)
* allow user specific encoding / decoding of binary
# v0.3.1
* compatible with Elixir v0.15.1
# v0.3
* compatible with Elixir v0.14.1


test/bson_test.exs
Code.require_file "test_helper.exs", __DIR__

defmodule Bson.Test do
  use ExUnit.Case

  doctest Bson
  doctest Bson.ObjectId
  doctest Bson.UTC
  doctest Bson.Decoder
  doctest Bson.Encoder.Protocol.Float
  doctest Bson.Encoder.Protocol.Integer
  doctest Bson.Encoder.Protocol.Atom
  doctest Bson.Encoder.Protocol.Bson.Regex
  doctest Bson.Encoder.Protocol.Bson.ObjectId
  doctest Bson.Encoder.Protocol.Bson.JS
  doctest Bson.Encoder.Protocol.Bson.Bin
  doctest Bson.Encoder.Protocol.Bson.Timestamp
  doctest Bson.Encoder.Protocol.BitString
  doctest Bson.Encoder.Protocol.Bson.UTC
  doctest Bson.Encoder.Protocol.List
  doctest Bson.Encoder.Protocol.Map

end


test/bsonjson_test.exs
Code.require_file "test_helper.exs", __DIR__

defmodule BsonJson.Test do
  use ExUnit.Case

  test "Stringify" do
  	assert {"{}", ""} = BsonJson.stringify(Bson.encode(%{}))
  	assert {"{\"a\":1}", ""} = BsonJson.stringify(Bson.encode(%{a: 1}))
  	assert {"{\"a\":1.1}", ""} = BsonJson.stringify(Bson.encode(%{a: 1.1}))
  	assert {"{\"a\":\"ab\"}", ""} = BsonJson.stringify(Bson.encode(%{a: "ab"}))
  	assert {"{\"a\":\"a\",\"b\":\"b\"}", ""} = BsonJson.stringify(Bson.encode(%{a: "a", b: "b"}))
  	assert {"{\"a\":[1,2]}", ""} = BsonJson.stringify(Bson.encode(%{a: [1,2]}))
  end

  test "ObjectId" do
    term = %Bson.ObjectId{oid: <<82, 224, 252, 230, 0, 0, 2, 0, 3, 0, 0, 4>>}
    bson = Bson.encode(%{"0": term})

    assert {"{\"0\":\"52e0fce60000020003000004\"}", ""} == BsonJson.stringify(bson)
  end

end

mix.exs
defmodule Bson.Mixfile do
  use Mix.Project

  def project do
    [ app: :bson,
      name: "bson",
      version: "0.4.4",
      elixir: "~> 1.0.0",
      description: "BSON implementation for Elixir",
      source_url: "https://github.com/checkiz/elixir-bson",
      deps: deps(Mix.env),
      package: package,
      docs: &docs/0 ]
  end

  # Configuration for the OTP application
  def application do
    []
  end

  # Returns the list of dependencies in the format:
  defp deps(:docs) do
    [
      {:ex_doc, ">= 0.0.0" },
      {:earmark, ">= 0.0.0"}
    ]
  end
  defp deps(_), do: []

  defp docs do
    [ #readme: true,
      #main: "README",
      source_ref: System.cmd("git", ["rev-parse", "--verify", "--quiet", "HEAD"])|>elem(0) ]
  end

  defp package do
    [ contributors: ["jerp"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/checkiz/elixir-bson",
        "Documentation" => "http://hexdocs.pm/bson/"
      } ]
  end
end


lib/bsonjson.ex
defmodule BsonJson do
  @moduledoc """
  Converts a Bson document into a JSON document.

  """

  @doc """
  Returns a json representation of set of Bson documents
  transcodeing the following element type:

  * int32 -> number
  * int64 -> number (capped at js maximum)
  * float -> number
  * string -> string (utf8)
  * document -> object
  * array document -> array
  * objectId -> 24 character length hexadecimal string
  """
  def stringify(bson) do
    case document(bson) do
      {acc, rest} -> {acc|>List.flatten|>IO.iodata_to_binary, rest}
    end
  end
  defp int32(<<i::size(32)-signed-little, rest::binary>>), do: {to_string(i), rest}
  defp int64(<<i::size(64)-signed-little, rest::binary>>), do: {to_string(i), rest}

  defp float(<<0, 0, 0, 0, 0, 0, 248, 127, rest::binary>>), do: {"null", rest} #nan
  defp float(<<0, 0, 0, 0, 0, 0, 248, 255, rest::binary>>), do: {"null", rest} #nan
  defp float(<<0, 0, 0, 0, 0, 0, 240, 127, rest::binary>>), do: {"9007199254740992", rest}  #+inf
  defp float(<<0, 0, 0, 0, 0, 0, 240, 255, rest::binary>>), do: {"-9007199254740992", rest} #-inf
  defp float(<<f::size(64)-float-little, rest::binary>>), do: {to_string(f), rest}

  defp string(<<l::size(32)-signed-little, rest::binary>>) do
    bitsize = (l-1)*8
    <<string::size(bitsize), 0, rest::binary>> = rest
    { [?", <<string::size(bitsize)>>, ?"],
      rest }
  end
  defp objectid(<<oid::96, rest::binary>>) do
    {<<?">> <> (for << <<b::size(4)>> <- <<oid::size(96)>> >>, into: <<>> do
          <<Integer.to_string(b,16)::binary>>
        end |> String.downcase) <> <<?">>, rest}
  end

  defp document(<<l::size(32)-signed-little, rest::binary>>) do
    bitsize = (l-5)*8
    <<bsondoc::size(bitsize), 0, rest::binary>> = rest
    { document(<<bsondoc::size(bitsize)>>, '', '{'), rest}
  end
  defp document("", _, acc), do: Enum.reverse([?}|acc])
  defp document(<<head, rest::binary>>, prefix, acc) do
    {el_name, rest} = peek_cstring(rest, [])
    {el_value, rest} = element(head, rest)
    document(rest, ?,, [el_value, ?:, ?", el_name, ?", prefix | acc])
  end

  defp array(<<l::size(32)-signed-little, rest::binary>>) do
    bitsize = (l-5)*8
    <<bsondoc::size(bitsize), 0, rest::binary>> = rest
    { array(<<bsondoc::size(bitsize)>>, '', [?[]), rest}
  end
  defp array("", _, acc), do: Enum.reverse([?]|acc])
  defp array(<<head, rest::binary>>, prefix, acc) do
    {_, rest} = peek_cstring(rest, [])
    {el_value, rest} = element(head, rest)
    array(rest, ?,, [el_value, prefix | acc])
  end

  defp element(head, bson) do
    case head do
      0x01 -> float(bson)
      0x02 -> string(bson)
      0x03 -> document(bson)
      0x04 -> array(bson)
      0x07 -> objectid(bson)
      0x10 -> int32(bson)
      0x12 -> int64(bson)
    end
  end

  defp peek_cstring(<<0, rest::binary>>, acc), do: {acc|>Enum.reverse|>IO.iodata_to_binary, rest}
  defp peek_cstring(<<c, rest::binary>>, acc), do: peek_cstring(rest, [c|acc])
  defp peek_cstring("", _acc), do: raise "bson corrupted: expecting cstring end mark"
end


lib/bson_encoder.ex
defmodule Bson.Encoder do
  defprotocol Protocol do
    @moduledoc """
    `Bson.Encoder.Protocol` protocol defines Bson encoding according to Elxir terms and some Bson predefined structs (see `Bson`).

    List of the protocol implementations:

    * `Map` - Encodes a map into a document
    * `HasDict` - Encodes a HashDict into a document
    * `Keyword` - Encodes a Keyword into a document
    *  `List` - Encodes a list of key-alue pairs into a document otherwize encode list into array
    * `Integer` - Encodes integer in 32 or 64 bits
    * `Float` - Encodes float in 64 bits
    * `Atom` - Encodes special atom (`false`, `true`, `nil`,
    `:nan`, `:+inf`, `:-inf`, `MIN_KEY` and `MAX_KEY`) in appropriate format
    others in special type Symbol
    * `BitString` - as binary string
    * `Bson.Regex' - see specs
    * `Bson.ObjectId' - see specs
    * `Bson.JS' - see specs
    * `Bson.Bin' - see specs
    * `Bson.Timestamp  ' - see specs
    """

    @doc """
    Returns a binary representing a term in Bson format
    """
    def encode(term)
  end
  defmodule Error do
    @moduledoc """
    Container for error messages

    * `what` has triggerred the error
    * `acc` contains what was already decoded for this term (ie the size of a string when the string itself could not be decoded)
    * `term` that failed to be encoded
    """
    defstruct [what: nil, acc: [], term: nil]
    defimpl Inspect, for: Error do
      def inspect(e,_), do: inspect([what: e.what, term: e.term, acc: e.acc])
    end
  end

  @doc """
  Creates a document using a collection of element, this is, a key-value pair
  """
  def document(element_list) do
    case Enumerable.reduce(element_list, {:cont, []},
      fn({key, value}, acc) when is_binary(key) -> accumulate_elist(key, value, acc)
        ({key, value}, acc) when is_atom(key) -> accumulate_elist(Atom.to_string(key), value, acc)
        (element, acc) -> {:halt, %Error{what: [:element], term: element, acc: acc |> Enum.reverse}}
      end) do
      {:halted, error} -> error
      {:done, acc} ->
        acc |> Enum.reverse |> IO.iodata_to_binary |> wrap_document
    end
  end

  defimpl Protocol, for: Integer do
    @doc """
    iex> Bson.Encoder.Protocol.encode(2)
    {<<16>>, <<2, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode(-2)
    {<<16>>, <<254, 255, 255, 255>>}
    iex> Bson.Encoder.Protocol.encode -0x80000001
    {<<18>>, <<255, 255, 255, 127, 255, 255, 255, 255>>}

    iex> Bson.Encoder.Protocol.encode 0x8000000000000001
    %Bson.Encoder.Error{what: [Integer], term: 0x8000000000000001}
    """
    def encode(i) when -0x80000000 <= i and i <= 0x80000000, do: {<<0x10>>, <<i::32-signed-little>>}
    def encode(i) when -0x8000000000000000 <= i and i <= 0x8000000000000000, do: {<<0x12>>, <<i::64-signed-little>>}
    def encode(i), do: %Error{what: [Integer], term: i}
  end

  defimpl Protocol, for: Float do
    @doc """
    iex> Bson.Encoder.Protocol.encode(1.1)
    {<<1>>, <<154, 153, 153, 153, 153, 153, 241, 63>>}
    """
    def encode(f), do: {<<0x01>>, <<(f)::size(64)-float-little>>}
  end

  defimpl Protocol, for: Atom do
    @doc """
    iex> Bson.Encoder.Protocol.encode(true)
    {<<8>>, <<1>>}
    iex> Bson.Encoder.Protocol.encode(nil)
    {<<10>>, <<>>}
    iex> Bson.Encoder.Protocol.encode(:max_key)
    {<<127>>, <<>>}
    iex> Bson.Encoder.Protocol.encode(:min_key)
    {<<255>>, <<>>}
    iex> Bson.Encoder.Protocol.encode(:nan)
    {<<1>>, <<0, 0, 0, 0, 0, 0, 248, 127>>}
    iex> Bson.Encoder.Protocol.encode(:'+inf')
    {<<1>>, <<0, 0, 0, 0, 0, 0, 240, 127>>}
    iex> Bson.Encoder.Protocol.encode(:'-inf')
    {<<1>>, <<0, 0, 0, 0, 0, 0, 240, 255>>}
    iex> Bson.Encoder.Protocol.encode(:atom)
    {<<14>>, [<<5, 0, 0, 0>>, "atom", <<0>>]}

    """
    # predefind Bson value
    def encode(false),   do: {<<0x08>>, <<0x00>>}
    def encode(true),    do: {<<0x08>>, <<0x01>>}
    def encode(nil),     do: {<<0x0a>>, <<>>}
    def encode(:nan),    do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 248, 127>>}
    def encode(:'+inf'), do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 240, 127>>}
    def encode(:'-inf'), do: {<<0x01>>, <<0, 0, 0, 0, 0, 0, 240, 255>>}
    def encode(:min_key), do: {<<0xff>>, <<>>}
    def encode(:max_key), do: {<<0x7f>>, <<>>}
    # other Elixir atom are encoded like strings ()
    def encode(atom),    do: {<<0x0e>>, (atom |> Atom.to_string |> Bson.Encoder.wrap_string)}
  end

  defimpl Protocol, for: Bson.UTC do
    @doc """
    iex> Bson.Encoder.Protocol.encode(Bson.UTC.from_now({1390, 324703, 518471}))
    {<<9>>, <<30, 97, 207, 181, 67, 1, 0, 0>>}
    """
    def encode(%Bson.UTC{ms: ms}) when is_integer(ms), do: {<<0x09>>, <<ms::64-little-signed>>}
    def encode(utc), do: %Error{what: [Bson.UTC], term: utc}
  end

  defimpl Protocol, for: Bson.Regex do
    @doc """
    iex> Bson.Encoder.Protocol.encode(%Bson.Regex{pattern: "p", opts: "i"})
    {<<11>>, ["p", <<0>>, "i", <<0>>]}
    """
    def encode(%Bson.Regex{pattern: p, opts: o}) when is_binary(p) and is_binary(o), do: {<<0x0b>>, [p, <<0x00>>, o, <<0x00>>]}
    def encode(regex), do: %Error{what: [Bson.Regex], term: regex}
  end

  defimpl Protocol, for: Bson.ObjectId do
    @doc """
    iex> Bson.Encoder.Protocol.encode(%Bson.ObjectId{oid: <<0xFF>>})
    {<<0x07>>, <<255>>}

    iex> Bson.Encoder.Protocol.encode(%Bson.ObjectId{oid: 123})
    %Bson.Encoder.Error{what: [Bson.ObjectId], term: %Bson.ObjectId{oid: 123}}
    """
    def encode(%Bson.ObjectId{oid: oid}) when is_binary(oid), do: {<<0x07>>, oid}
    def encode(oid), do: %Error{what: [Bson.ObjectId], term: oid}
  end

  defimpl Protocol, for: Bson.JS do
    @doc """
    iex> Bson.Encoder.Protocol.encode(%Bson.JS{code: "1+1;"})
    {<<13>>, [<<5, 0, 0, 0>>, "1+1;", <<0>>]}
    iex> Bson.Encoder.Protocol.encode(%Bson.JS{code: "1+1;", scope: %{a: 0, b: "c"}})
    {<<15>>, <<34, 0, 0, 0, 5, 0, 0, 0, 49, 43, 49, 59, 0, 21, 0, 0, 0, 16, 97, 0, 0, 0, 0, 0, 2, 98, 0, 2, 0, 0, 0, 99, 0, 0>>}
    """
    def encode(%Bson.JS{code: js, scope: nil}) when is_binary(js) do
      {<<0x0d>>, Bson.Encoder.wrap_string(js)}
    end
    def encode(%Bson.JS{code: js, scope: ctx}) when is_binary(js) and is_map(ctx) do
      case Bson.Encoder.document(ctx) do
        %Error{}=error -> %Error{error|what: {:js_context, error.what}}
        ctxBin ->
          {<<0x0f>>, [Bson.Encoder.wrap_string(js), ctxBin] |> IO.iodata_to_binary |> js_ctx}
      end
    end
    def encode(js), do: %Error{what: [Bson.JS], term: js}

    defp js_ctx(jsctx), do: <<(byte_size(jsctx)+4)::32-little-signed, jsctx::binary>>
  end

  defimpl Protocol, for: Bson.Bin do
    @doc """
    iex> Bson.Encoder.Protocol.encode(%Bson.Bin{bin: "e", subtype: Bson.Bin.subtyx(:user)})
    {<<5>>,[<<1, 0, 0, 0>>, 128, "e"]}
    """
    def encode(%Bson.Bin{bin: bin, subtype: subtype}), do: encode(bin, subtype)
    def encode(bin, subtype)
      when is_binary(bin) and is_integer(subtype),
      do:  {<<0x05>>, [<<byte_size(bin)::32-little-signed>>, subtype, bin]}
    def encode(bin, subtype), do: %Error{what: [Bson.Bin], term: {bin, subtype}}
  end

  defimpl Protocol, for: Bson.Timestamp do
    @doc """
    iex> Bson.Encoder.Protocol.encode(%Bson.Timestamp{inc: 1, ts: 2})
    {<<17>>,<<1, 0, 0, 0, 2, 0, 0, 0>>}
    """
    def encode(%Bson.Timestamp{inc: i, ts: t})
      when is_integer(i) and -0x80000000 <= i and i <= 0x80000000
       and is_integer(t) and -0x80000000 <= t and t <= 0x80000000,
      do: {<<0x11>>, <<i::32-signed-little, t::32-signed-little>>}
      def encode(ts), do: %Error{what: [Bson.Timestamp], term: ts}
  end

  defimpl Protocol, for: BitString do
    @doc """
    iex> Bson.Encoder.Protocol.encode("a")
    {<<2>>, [<<2, 0, 0, 0>>, "a", <<0>>]}
    """
    def encode(s) when is_binary(s),  do: {<<0x02>>, Bson.Encoder.wrap_string(s)}
    def encode(bits), do: %Error{what: [BitString], term: bits}
  end

  defimpl Protocol, for: List do
    @doc """
    iex> Bson.Encoder.Protocol.encode([])
    {<<4>>,<<5, 0, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode([2, 3])
    {<<4>>,<<19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode([1,[nil]])
    {<<4>>,<<23, 0, 0, 0, 16, 48, 0, 1, 0, 0, 0, 4, 49, 0, 8, 0, 0, 0, 10, 48, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode([1,[2, 3]])
    {<<4>>,<<34, 0, 0, 0, 16, 48, 0, 1, 0, 0, 0, 4, 49, 0, 19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0, 0>>}

    # Keyword and list of key-value pairs
    iex> Bson.Encoder.Protocol.encode([a: "r"])
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode([{"a", "s"}])
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 115, 0, 0>>}

    iex> Bson.Encoder.Protocol.encode([{"a", "s"}, {:b, "r"}, 1, 2])
    %Bson.Encoder.Error{
      term: 1,
      what: [:element],
      acc: [[<<2>>, "a", <<0>>, [<<2, 0, 0, 0>>, "s", <<0>>]],
            [<<2>>, "b", <<0>>, [<<2, 0, 0, 0>>, "r", <<0>>]]]}

    iex> Bson.Encoder.Protocol.encode([2, 3, ])
    {<<4>>,<<19, 0, 0, 0, 16, 48, 0, 2, 0, 0, 0, 16, 49, 0, 3, 0, 0, 0, 0>>}

    """
    def encode([{k, _}|_]=elist) when is_atom(k) or is_binary(k) do
      case Bson.Encoder.document(elist) do
        %Error{}=error -> error
        encoded_elist -> {<<0x03>>, encoded_elist}
      end
    end
    def encode(list) do
      case Bson.Encoder.array(list) do
        %Error{}=error -> error
        encoded_list -> {<<0x04>>, encoded_list}
      end
    end
  end

  defimpl Protocol, for: [Map, HashDict, Keyword] do
    @doc """
    # Map
    iex> Bson.Encoder.Protocol.encode(%{})
    {<<3>>,<<5, 0, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode(%{a: "r"})
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode(%{a: 1, b: 5})
    {<<3>>,<<19, 0, 0, 0, 16, 97, 0, 1, 0, 0, 0, 16, 98, 0, 5, 0, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode(%{a: 1, b: %{c: 3}})
    {<<3>>,<<27, 0, 0, 0, 16, 97, 0, 1, 0, 0, 0, 3, 98, 0, 12, 0, 0, 0, 16, 99, 0, 3, 0, 0, 0, 0, 0>>}

    # HashDict
    iex> Bson.Encoder.Protocol.encode(%HashDict{})
    {<<3>>,<<5, 0, 0, 0, 0>>}
    iex> Bson.Encoder.Protocol.encode(HashDict.put(%HashDict{}, :a, "r"))
    {<<3>>,<<14, 0, 0, 0, 2, 97, 0, 2, 0, 0, 0, 114, 0, 0>>}

    iex> Bson.Encoder.Protocol.encode(%{a: "va", b: "vb", u: %Bson.UTC{ms: "e"}})
    %Bson.Encoder.Error{
      what: ["u", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
             [<<2>>, "b", <<0>>, [<<3, 0, 0, 0>>, "vb", <<0>>]]]]}
    iex> Bson.Encoder.Protocol.encode([1, 2, %Bson.UTC{ms: "e"}])
    %Bson.Encoder.Error{
      what: ["2", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<16>>, "0", <<0>>, <<1, 0, 0, 0>>],
             [<<16>>, "1", <<0>>, <<2, 0, 0, 0>>]]]}
    iex> Bson.Encoder.Protocol.encode(%{a: "va", b: "vb", c: %{c1: "vc1", cu: %Bson.UTC{ms: "e"}}})
    %Bson.Encoder.Error{
      what: ["c", "cu", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
            [<<2>>, "b", <<0>>,  [<<3, 0, 0, 0>>, "vb", <<0>>]]],
           [[<<2>>, "c1", <<0>>, [<<4, 0, 0, 0>>, "vc1", <<0>>]]]]}
    iex> Bson.Encoder.Protocol.encode(%{a: "va", b: "vb", c: ["c0", %Bson.UTC{ms: "e"}]})
    %Bson.Encoder.Error{
      what: ["c", "1", Bson.UTC],
      term: %Bson.UTC{ms: "e"},
      acc: [[[<<2>>, "a", <<0>>, [<<3, 0, 0, 0>>, "va", <<0>>]],
            [<<2>>, "b", <<0>>, [<<3, 0, 0, 0>>, "vb", <<0>>]]],
           [[<<2>>, "0", <<0>>, [<<3, 0, 0, 0>>, "c0", <<0>>]]]]}

    """
    def encode(dict) do
      case Bson.Encoder.document(dict) do
        %Error{}=error -> error
        encoded_dict -> {<<0x03>>, encoded_dict}
      end
    end
  end

  @doc """
  Creates a document for an array (list of items)
  """
  def array(item_list) do
    case Enumerable.reduce(item_list, {:cont, {[], 0}},
      fn(item, {acc, i}) ->
        case accumulate_elist(Integer.to_string(i), item, acc) do
          {:cont, acc} -> {:cont, {acc, i+1}}
          {:halt, error} -> {:halt, error}
        end
      end) do
      {:halted, error} -> error
      {:done, {bufferAcc, _}} ->
        bufferAcc |> Enum.reverse |> IO.iodata_to_binary |> wrap_document
    end
  end

  @doc """
  Wraps a bson document with size and trailing null character
  """
  def wrap_document(elist), do: <<(byte_size(elist)+5)::32-little-signed>> <> elist <> <<0x00>>

  @doc """
  Wraps a bson document with size and trailing null character
  """
  def wrap_string(string), do: [<<(byte_size(string)+1)::32-little-signed>>, string, <<0x00>>]

  @doc """
  Accumulate element in an element list
  """
  def accumulate_elist(name, value, elist) do
    case element(name, value) do
      %Error{}=error -> {:halt, %Error{error|acc: [Enum.reverse(elist)|error.acc]}}
      encoded_element -> {:cont, [encoded_element | elist]}
    end
  end

  @doc """
  Returns encoded element using its name and value
  """
  def element(name, value) do
    case Bson.Encoder.Protocol.encode(value) do
      %Error{}=error -> %Error{error|what: [name|error.what]}
      {kind, encoded_value} -> [kind, name, <<0x00>>, encoded_value]
    end
  end
end
